victim-loss-identifier/src/agent.ts

```diff
@@ -1,16 +1,255 @@
+                erc721: {
+                  [stolenTokenAddress]: {
+                    // TODO: Would this be an issue if it
+                    // overwrites the existing name and symbol?
+                    tokenName: stolenTokenName,
+                    tokenSymbol: stolenTokenSymbol,
+                    tokenIds: [],
+                    tokenTotalUsdValue: BigNumber.from("0"),
+                  },
+                },
+              },
+            },
+          },
+        };
+
+        scammersCurrentlyMonitored[scammerAddress].victims![victimAddress].totalUsdValueAcrossAllTokens!.add(
+          utils.parseEther(nftCollectionFloorPrice.toString())
+        );
+        scammersCurrentlyMonitored[scammerAddress].victims![victimAddress].transactions![exploitTxnHash].erc721![
+          stolenTokenAddress
+        ].tokenIds!.push(stolenTokenId);
+        scammersCurrentlyMonitored[scammerAddress].victims![victimAddress].transactions![exploitTxnHash].erc721![
+          stolenTokenAddress
+        ].tokenTotalUsdValue!.add(utils.parseEther(nftCollectionFloorPrice.toString()));
+
+        findings.push(
+          createFraudNftOrderFinding(
+            victimAddress,
+            scammerAddress,
+            scammersCurrentlyMonitored[scammerAddress].firstAlertIdAppearance,
+            scammersCurrentlyMonitored[scammerAddress].victims![victimAddress].totalUsdValueAcrossAllTokens!,
+            stolenTokenName,
+            stolenTokenAddress,
+            stolenTokenId,
+            scammersCurrentlyMonitored[scammerAddress].victims![victimAddress].transactions![exploitTxnHash].erc721![
+              stolenTokenAddress
+            ].tokenTotalUsdValue!,
+            exploitTxnHash,
+            nftCollectionFloorPrice
+          )
+        );
+      }
+    })
+  );
+
+  return findings;
+}
+
+export function provideInitialize(provider: providers.Provider, apiKeyFetcher: () => Promise<string>): Initialize {
+  return async () => {
+    chainId = (await provider.getNetwork()).chainId;
+    apiKey = await apiKeyFetcher();
+
+    alertConfig: {
+      subscriptions: [
+        {
+          botId: SCAM_DETECTOR_BOT_ID,
+          alertIds: SCAM_DETECTOR_ALERT_IDS,
+        },
+      ];
+    }
+  };
+}
+
+export function provideHandleAlert(
+  provider: providers.Provider,
+  getErc721TransfersInvolvingScammer: (
+    scammerAddress: string,
+    transferOccuranceTimeWindow: number,
+    apiKey: string
+  ) => Promise<any[]>,
+  fetchNftCollectionFloorPrice: () => BigNumber
+): HandleAlert {
+  return async (alertEvent: AlertEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    const scammerAddress = alertEvent.alert.metadata["scammer_addresses"];
+    if (!Object.keys(scammersCurrentlyMonitored).includes(scammerAddress)) {
+      scammersCurrentlyMonitored[scammerAddress] = {
+        firstAlertIdAppearance: alertEvent.alertId!,
+        mostRecentActivityByBlockNumber: alertEvent.blockNumber!,
```

Comment:
Should this be updated even if the address is already monitored?

Reply:
You mean that `mostRecentActivityByBlockNumber` should be updated regardless of whether it is already monitored, right?

If so, that logic to handle scammers that are already included is in the block handler. 🙂 

(If that's what you mean)